cpp/src/arrow/compute/registry.cc

```diff
@@ -276,6 +276,7 @@ static std::unique_ptr<FunctionRegistry> CreateBuiltInRegistry() {
   // Register core kernels
   RegisterScalarCast(registry.get());
   RegisterDictionaryDecode(registry.get());
+  RegisterDictionaryCompaction(registry.get());
```

Comment:
Sorry, I have no consideration on these.

> should that be done in the #if ARROW_COMPUTE block below?

Have updated

> Granted, I'm pretty sure doing that would also necessitate moving the tests out of vector_hash_test.cc since it's built/run unconditionally.

Sure, actully, I have no idea where to place the codes. Do we need to create a new file `vector_dictionary.cc` and `vector_dictionary_test.cc`?

Reply:
> Sorry, I have no consideration on these.

Understandable. The whole situation isn't really ideal at the moment...

> Do we need to create new files `vector_dictionary.cc` and `vector_dictionary_test.cc`?

You _could_ get away with only moving the tests, but creating `vector_dictionary.cc` would be better. Just make sure it's added to the appropriate section in the CMakeLists: https://github.com/R-JunmingChen/arrow/blob/46959c21d2ca6a2bdd8026e407c90a34b907fdb6/cpp/src/arrow/CMakeLists.txt#L434

Reply:
I add `vector_dictionary.cc` and `vector_dictionary_test.cc`.